### ISSUE TITLE
Modularity of components

### ISSUE THREAD

**Issue by soniktrooth** [kalamuna/kalastatic#320]
Currently when we write our Sass files we do something like the following: 

```
%header {
  background: red;
  padding: 2em;
}

.header {
  @extend %header;
}
```

This is great and it works most of the time, however for true modularity I think we need to split these out. 

I came across a situation today on GBZ where I want to pull in component styling from one theme to the other and the current set up as described above doesn't work because I want to style the header with this new component within a special set of circumstances. 

I realize there's extra overhead to splitting up the definition and application of the styling but I wanted to have some discussion around it. It's probably not going to be required in all sites but we should have a way to achieve it on more complex sites like GBZ where it would be beneficial to share components across themes.

Thoughts @thiagodemellobueno ?

**Comment by RobLoach** [comment]
Definition and implementation should be separate. Good find, Josh.

**Comment by thiagodemellobueno** [comment]
A++


Could you please elaborate on:
"I want to style the header with this new component within a special set of circumstances."

I imagine unpacking this bit will aid in our finding success here.

On July 11, 2016 8:29:16 PM EDT, Josh <notifications@github.com> wrote:
>Currently when we write our Sass files we do something like the
>following: 
>
>```
>%header {
>  background: red;
>  padding: 2em;
>}
>
>.header {
>  @extend %header;
>}
>```
>
>This is great and it works most of the time, however for true
>modularity I think we need to split these out. 
>
>I came across a situation today on GBZ where I want to pull in
>component styling from one theme to the other and the current set up as
>described above doesn't work because I want to style the header with
>this new component within a special set of circumstances. 
>
>I realize there's extra overhead to splitting up the definition and
>application of the styling but I wanted to have some discussion around
>it. It's probably not going to be required in all sites but we should
>have a way to achieve it on more complex sites like GBZ where it would
>be beneficial to share components across themes.
>
>Thoughts @thiagodemellobueno ?
>
>
>
>---
>You are receiving this because you were assigned.
>Reply to this email directly or view it on GitHub:
>https://github.com/kalamuna/kalastatic/issues/320

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by soniktrooth** [comment]
In the case I was referring to I wanted to apply the new styling to the header only when a specific class appeared on the body. Whereas, importing a file that contains the definition and application means that I would get this new styling applied to the header always.

**Comment by thiagodemellobueno** [comment]
@soniktrooth 
Do you think simply having all placeholder / components definitions in "components" and moving all the "site specific styles" to src/styles/{Same structure forcomponent directories}/{_component.scss} which holds all the "bindings."

It's sort of the first thing to come to mind, but I cant seem to poke a hole through it, and keeps our src/component things very DRY and reusable.

@soniktrooth, @RobLoach, can you envision any downsides to this approach?

**Comment by thiagodemellobueno** [comment]
So based on our convo this standup...  
src/components has reusable %placeholders
but a specific theme might just have
src/styles/organism_name/_organismName.scss where the  atoms and molecules extend selectors....

**Comment by thiagodemellobueno** [comment]
This maybe interesting in this context (maybe not) https://www.smashingmagazine.com/2013/08/other-interface-atomic-design-sass/

**Comment by thiagodemellobueno** [comment]
Recent experience has me thinking about this one lately.

Thought experiment... 
Only placeholders in the components directory.

Extending happens in src/styles we define modular abstractions such content types or views or pages (or even some molecules and organisms) where we apply the real world selector approach.... 

hypothetical example
```/src/styles/modal/modal.scss```

```
.modal .view-mode-gbz_event_speaker_details {

  @extend %speaker-modal; // defined as an organism in src/components/organisms/speaker_modal/

  .title {
      @extend %section-title; // defined as an atom in src/components/atoms/section_title
   }

   .field-name-name-field-field-name_namename-_speaker-image {
       @extend %speaker-photo; // defined as an atom in src/components/atoms/speaker_photo
   }
}
```

I'm not endorsing this proposal yet, just wanted to document the questions it may rase, and validate.

@soniktrooth what are your thoughts?

**Comment by soniktrooth** [comment]
Technically, I think your example is solid @thiagodemellobueno and I love the flexibility.

My only fear is this will make the style application less discoverable/more mysterious because it's living in another file(s). 

We could mandate a comment in the declaration file when a component is usedbut this is easily forgotten and hard to maintain. Maybe some extra build/grunt/node fanciness from @thiagodemellobueno or @RobLoach could trawl the repo and add a comment in the component file as to where in the project the component is being used—just a thought.

**Comment by thiagodemellobueno** [comment]
eenteresting

**Comment by thiagodemellobueno** [comment]
@soniktrooth how has your thinking evolved in regard to this?

**Comment by soniktrooth** [comment]
@thiagodemellobueno yeah, I think the practicalities probably outweigh the purity for 90% of projects. At least we have talked about it and can cross the bridge at some point when we know it's needed.
